src/rebuild/srv.c

```diff
@@ -926,13 +973,39 @@ rebuild_try_merge_tgts(const uuid_t pool_uuid, uint32_t map_ver,
 {
 	struct rebuild_task *task;
 	struct rebuild_task *found = NULL;
-	int		    rc;
+	int rc;
+	uint32_t highest_queued_map_ver = -1;
 
-	d_list_for_each_entry(task, &rebuild_gst.rg_queue_list,
-			      dst_list) {
-		/* Only merge tasks that match both pool AND operation */
+	/* Loop over the queued tasks, looking for existing queued tasks for
+	 * this pool that match the UUID.
+	 *
+	 * For all such matching tasks, find the highest queued map_ver
+	 */
+	d_list_for_each_entry(task, &rebuild_gst.rg_queue_list, dst_list) {
+		if (uuid_compare(task->dst_pool_uuid, pool_uuid) == 0) {
+			/* This task is for the same UUID */
+			if (highest_queued_map_ver < task->dst_map_ver) {
+				highest_queued_map_ver = task->dst_map_ver;
+			}
+		}
+	}
+
+	/* Loop over the queued tasks again, looking for one that matches this
+	 * pool AND this operation AND is the highest found version.
+	 *
+	 * This will be safe to merge to - because there are no other operation
+	 * types that were queued after that (they would have a higher version).
+	 *
+	 * A task with a lower map version than the highest found will not be
+	 * safe to merge to, even if it's the same type, because this will cause
+	 * rebuild to essentially skip the intermediary different-type step
+	 * because the rebuild version is set to the task map version after
+	 * rebuild is complete.
+	 */
+	d_list_for_each_entry(task, &rebuild_gst.rg_queue_list, dst_list) {
 		if (uuid_compare(task->dst_pool_uuid, pool_uuid) == 0
-		    && task->dst_rebuild_op == rebuild_op) {
+		    && task->dst_rebuild_op == rebuild_op
+		    && task->dst_map_ver == highest_queued_map_ver) {
```

Comment:
you could do this in one loop by change the "if" 
```
do_list_for_each_entry(..) {
    if (uuid_compare(...) != 0)
           continue;
     if (task->dst_rebuild_op != rebuild_op)
            found = NULL;
     else
            found = task;
 }
```
Now you don't need to previous loop correct?

Reply:
Good observation. I've refactored this as you suggest, which should work.